cores/arduino/SoftwareSerial.cpp

```diff
@@ -0,0 +1,420 @@
+uint32_t SoftwareSerial::tx_buffer = 0;
+int32_t SoftwareSerial::tx_bit_cnt = 0;
+uint32_t SoftwareSerial::rx_buffer = 0;
+int32_t SoftwareSerial::rx_bit_cnt = -1;
+uint32_t SoftwareSerial::cur_speed = 0;
+
+//
+// Private methods
+//
+
+void SoftwareSerial::setSpeed(uint32_t speed)
+{
+  if (speed != cur_speed) {
+    timer.pause();
+    if (speed != 0) {
+      // Disable the timer
+      uint32_t clock_rate, cmp_value;
+      // Get timer clock
+      clock_rate = timer.getTimerClkFreq();
+      int pre = 1;
+      // Calculate prescale an compare value
+      do {
+        cmp_value = clock_rate / (speed * OVERSAMPLE);
+        if (cmp_value >= UINT16_MAX) {
+          clock_rate = clock_rate / 2;
+          pre *= 2;
+        }
+      } while (cmp_value >= UINT16_MAX);
+      timer.setPrescaleFactor(pre);
+      timer.setOverflow(cmp_value);
+      timer.setCount(0);
+      timer.attachInterrupt(&handleInterrupt);
+      timer.resume();
+    } else {
+      timer.detachInterrupt();
+    }
+    cur_speed = speed;
+  }
+}
+
+// This function sets the current object as the "listening"
+// one and returns true if it replaces another
+bool SoftwareSerial::listen()
+{
+  if (_receivePin >= 0) {
+    // wait for any transmit to complete as we may change speed
+    while (active_out)
+      ;
+    if (active_listener) {
+      active_listener->stopListening();
+    }
+    rx_tick_cnt = 1;
+    rx_bit_cnt = -1;
+    setSpeed(_speed);
+    active_listener = this;
+    if (!_half_duplex) {
+      active_in = this;
+    }
+    return true;
+  }
+  return false;
+}
+
+// Stop listening. Returns true if we were actually listening.
+bool SoftwareSerial::stopListening()
+{
+  if (active_listener == this) {
+    // wait for any output to complete
+    while (active_out)
+      ;
+    if (_half_duplex) {
+      setRXTX(false);
+    }
+    active_listener = nullptr;
+    active_in = nullptr;
+    // turn off ints
+    setSpeed(0);
+    return true;
+  }
+  return false;
+}
+
+inline void SoftwareSerial::setTX()
+{
+  digitalWriteFast(digitalPinToPinName(_transmitPin), _inverse_logic ? LOW : HIGH);
+  pinMode(_transmitPin, OUTPUT);
+}
+
+inline void SoftwareSerial::setRX()
+{
+  if (_receivePin > 0) {
+    pinMode(_receivePin, _inverse_logic ? INPUT_PULLDOWN : INPUT_PULLUP); // pullup for normal logic!
+  }
+}
+
+inline void SoftwareSerial::setRXTX(bool input)
+{
+  if (_half_duplex) {
+    if (input) {
+      if (active_in != this) {
+        setRX();
+        rx_bit_cnt = -1;
+        rx_tick_cnt = 2;
+        active_in = this;
+      }
+    } else {
+      if (active_in == this) {
+        setTX();
+        active_in = nullptr;
+      }
+    }
+  }
+}
+
+inline void SoftwareSerial::send()
+{
+  if (--tx_tick_cnt <= 0) {
+    if (tx_bit_cnt++ < 10) {
+      // send data (including start and stop bits)
+      digitalWriteFast(digitalPinToPinName(_transmitPin), (tx_buffer & 1) ? ~_inverse_logic : _inverse_logic);
```

Comment:
In order to optimise/minimise Interrupt work:
1) Use STM32 Cube LL to set reset output pin (see example below)
2) save Port and pin number (LL format) in variable of the class thus convertion is done only once (in constructor or in setTX ...) . And in interrupt function recv() and send() you can directly call LL API without overhead.
3) Can we manage _inverse_logic  in write function ? so no work in interrupt fonctions, and also we can make the inversion can be done once for the whole byte (if (_inverse_logic)  tx_buffer = ~tx-buffer;) instead of once every bit.

Example LL:

_receivePinPort = digitalPinToPort(PinName);
_receivePinNumber = STM_LL_GPIO_PIN(digitalPinToPinName(PinName));
LL_GPIO_ReadInputPort(_receivePinPort);

_transmitPinPort = digitalPinToPort(PinName);
_transmitPinNumber = STM_LL_GPIO_PIN(digitalPinToPinName(PinName));
LL_GPIO_SetOutputPin(_transmitPinPort, _transmitPinNumber);  // Set High
LL_GPIO_ResetOutputPin(_transmitPinPort, _transmitPinNumber); // Set Low

Reply:
Ok, I will adapt the code.